librato_python_web/instrumentor/data/psycopg2.py

```diff
@@ -1,55 +1,88 @@
+        return self.__subject__.execute(*args, **kwargs)
+
+    def executemany(self, *args, **kwargs):
+        return self.__subject__.executemany(*args, **kwargs)
+
+    def fetchone(self, *args, **kwargs):
+        return self.__subject__.fetchone(*args, **kwargs)
+
+    def fetchmany(self, *args, **kwargs):
+        return self.__subject__.fetchmany(*args, **kwargs)
+
+    def fetchall(self, *args, **kwargs):
+        return self.__subject__.fetchall(*args, **kwargs)
+
+    def nextset(self, *args, **kwargs):
+        return self.__subject__.nextset(*args, **kwargs)
+
+
+class ConnWrapper(ObjectWrapper):
+    """ Wraps native connection class to permit instrumentation """
+    def cursor(self):
+        cursor = self.__subject__.cursor()
+        return WrappedCursor(cursor)
+
+
+def wrapped_connect(func, *args, **kwargs):
+    """ Returns a wrapped connection which intercepts cursor() """
+    conn = func(*args, **kwargs)
+    return ConnWrapper(conn)
 
 
 class Psycopg2Instrumentor(BaseInstrumentor):
     modules = {'psycopg2': ['connect'], 'psycopg2.extensions': ['connection', 'cursor']}
 
     def __init__(self):
         super(Psycopg2Instrumentor, self).__init__()
-        self.major_versions = [2]
 
     def run(self):
-        self.set_overridden(
-            {
-                'psycopg2': {
-                    'connect': {
-                        'returns': 'psycopg2.extensions.connection'
-                    }
-                },
-                'psycopg2.extensions.connection': {
-                    'cursor': {
-                        'returns': 'psycopg2.extensions.cursor'
-                    }
-                },
-            }
-        )
-
-        cursor_path = 'psycopg2.extensions.cursor.%s'
-
-        def func_name(e):
-            return e.split('(')[0]
-
-        def func_args(e):
-            return re.findall('[^(,)]+', e)[1:]
-
-        wrapped = {
-            cursor_path % func_name(m):
-                default_context_wrapper_factory('data.psycopg2.%s.' % func_name(m),
-                                                mapping={a: 1 for a in func_args(m)},
-                                                state='data.postgres', disable_if='model')
-            for m in 'callproc(resource),execute(resource),executemany(resource),fetchone,fetchmany,fetchall,'
-                     'nextset'.split(',')
-            }
-
-        wrapped['psycopg2.extensions.connection.cursor'] = wrap_returned_instance_decorator(
-            'psycopg2.extensions.cursor', wrapped
-        )
-
-        self.set_wrapped(
-            wrapped
-        )
+        # Instrument our wrapper connection class
+        meths = {
+                'librato_python_web.instrumentor.data.psycopg2.WrappedCursor.' + m:
+                get_complex_wrapper('data.psycopg2.%s.' % m, state='data.psycopg2', disable_if='model')
+
+                for m in ['callproc', 'execute', 'executemany', 'fetchone', 'fetchmany', 'fetchall', 'nextset']
```

Comment:
Is there a way to build this method list by introspecting the methods of `WrappedCursor`?